cpp/daal/src/algorithms/tsne/tsne_gradient_descent_avx512_impl.i

```diff
@@ -0,0 +1,326 @@
+
+        daal::threader_for(nBlocks, nBlocks, [&](IdxType iBlock) {
+            const IdxType iStart = iBlock * sizeOfBlock;
+            const IdxType iEnd   = services::internal::min<avx512, IdxType>(N, iStart + sizeOfBlock);
+            double * logLocal    = logTlsData.local();
+            if (!logLocal) return;
+
+            double * divLocal = divTlsData.local();
+            DAAL_CHECK_MALLOC_THR(divLocal);
+
+            xyType<double> row_point;
+            IdxType iCol, prefetch_index;
+            double y1d, y2d, sqDist, PQ;
+
+            for (IdxType iRow = iStart; iRow < iEnd; ++iRow)
+            {
+                size_t iSize      = 0;
+                mem._attr[iRow].x = 0.0;
+                mem._attr[iRow].y = 0.0;
+                row_point         = mem._pos[iRow];
+
+                if (!DivComp)
+                {
+                    IdxType start_index = row[iRow] - 1;
+                    IdxType range       = row[iRow + 1] - row[iRow];
+
+                    __m512d vec_1  = _mm512_set1_pd(1.0);
+                    __m256i vec_1i = _mm256_set1_epi32(1);
+
+                    __m512d vec_point_x = _mm512_set1_pd(row_point.x);
+                    __m512d vec_point_y = _mm512_set1_pd(row_point.y);
+
+                    __m512d vec_point_xs = _mm512_setzero_pd();
+                    __m512d vec_point_ys = _mm512_setzero_pd();
+
+                    for (IdxType i = 0; i < (range / 8) * 8; i += 8)
+                    {
+                        prefetch_index = start_index + i + prefetch_dist;
+                        if (prefetch_index < nnz) DAAL_PREFETCH_READ_T0(&mem._pos[col[prefetch_index] - 1]);
+
+                        __m256i vec_iCol = _mm256_slli_epi32(_mm256_sub_epi32(_mm256_loadu_epi32((__m256i *)&col[start_index + i]), vec_1i), 4);
+
+                        __m512d vec_point_xd = _mm512_sub_pd(vec_point_x, _mm512_i32gather_pd(vec_iCol, &mem._pos[0].x, 1));
+                        __m512d vec_point_yd = _mm512_sub_pd(vec_point_y, _mm512_i32gather_pd(vec_iCol, &mem._pos[0].y, 1));
+
+                        __m512d vec_pq = _mm512_div_pd(
+                            _mm512_loadu_pd((__m512d *)&val[start_index + i]),
+                            _mm512_add_pd(_mm512_fmadd_pd(vec_point_xd, vec_point_xd, _mm512_mul_pd(vec_point_yd, vec_point_yd)), vec_1));
+
+                        vec_point_xs = _mm512_fmadd_pd(vec_point_xd, vec_pq, vec_point_xs);
+                        vec_point_ys = _mm512_fmadd_pd(vec_point_yd, vec_pq, vec_point_ys);
+                    }
+
+                    mem._attr[iRow].x += _mm512_reduce_add_pd(vec_point_xs);
+                    mem._attr[iRow].y += _mm512_reduce_add_pd(vec_point_ys);
+
+                    for (IdxType i = (range / 8) * 8; i < range; ++i)
+                    {
+                        prefetch_index = start_index + i + prefetch_dist;
+                        if (prefetch_index < nnz) DAAL_PREFETCH_READ_T0(&mem._pos[col[prefetch_index] - 1]);
```

Comment:
```suggestion
                        if (prefetch_index < nnz) DAAL_PREFETCH_READ_T0((mem._pos + col[prefetch_index] - 1));
```

Reply:
I do not understand why the variant you propose is better? I think it does not affects the performance.